py/server/deephaven/pandas.py

```diff
@@ -279,3 +299,101 @@ def to_table(df: pd.DataFrame, cols: List[str] = None) -> Table:
+    dtypes.int32_array.j_type: np.dtype("i"),
+    dtypes.long_array.j_type: np.dtype("l"),
+    dtypes.float32_array.j_type: np.dtype("f"),
+    dtypes.double_array.j_type: np.dtype("d")
+}
+
+
+def _dh_null_conv_params(new_nv):
+    def decorator(fn):
+        """A decorator that replaces Deephaven null values in the parameters with the provided replacement value (
+        either numpy.nan or pandas.NA for now) before calling the decorated function.
+
+        Scalar type parameters must be type-hinted with precise numpy dtypes in order to be converted correctly,
+        e.g. np.int32, np.float64, etc. If the type-hint is not precise, the conversion will fail. if no type-hint is
+        provided, the conversion will skip the parameter.
+
+        Array type parameters don't need to be type-hinted, their types can be automatically detected for the conversion
+        be done correctly.
+        Note that:
+            if the new_nv is np.nan, the conversion automatically promotes arrays of integer types to array of float64.
+            if the new_nv is pd.NA, the conversion automatically promotes arrays to pd.Series of corresponding nullable
+            types.
+        """
+        fn_sig = _encode_signature(fn)
+
+        @wraps(fn)
+        def wrapper(*args, **kwargs):
+            converted_args = []
+            for arg, np_dtype_char in zip(args, fn_sig):
+                if np_dtype := _J_ARRAY_NP_TYPE_MAP.get(type(arg)):
+                    np_array = np.frombuffer(arg, dtype=np_dtype)
+                    dtype = dtypes.from_np_dtype(np_dtype)
+                    if new_nv is np.nan:
+                        dh_null = dtypes.PRIMITIVE_DTYPE_NULL_MAP.get(dtype)
+                        if dtype in (dtypes.float32, dtypes.float64):
+                            np_array = np.copy(np_array)
+                        else:
+                            np_array = np_array.astype(np.float64)
+                        np_array[np_array == dh_null] = new_nv
+                        converted_args.append(np_array)
+                    else:  # new_nv is pd.NA:
+                        pd_series = j_array_to_series(arg, dtype=dtype, conv_null=True)
+                        converted_args.append(pd_series)
+                else:
+                    try:
+                        np_dtype = np.dtype(np_dtype_char)
+                    except TypeError:
+                        converted_args.append(arg)
+                        continue
+                    dtype = dtypes.from_np_dtype(np_dtype)
+                    dh_null = dtypes.PRIMITIVE_DTYPE_NULL_MAP.get(dtype)
+                    if dh_null is not None:
+                        converted_args.append(new_nv if arg == dh_null else arg)
+                    else:
+                        converted_args.append(arg)
+
+            return fn(*converted_args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def dh_null_to_nan(fn):
```

Comment:
1. Should this function be in `numpy.py`?
2. Should this handle strings and map null to `""`?

Reply:
1. ideally yes, but that causes a cyclic import problem, we probably should just expose them in the top level so that we don't have this impression of misplacement
2. Nulls of Strings or non-primitive object types are returned as None via JPY.